File: src/gtfs.ts
/*
jvig - GTFS Viewer application written using Typescript & Electron
Copyright © 2020-2021 Mikołaj Kuranowski

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import * as fs from 'fs'

import { join } from 'path'
import { promisify } from 'util'

import * as pEvent from 'p-event'
import * as yauzl from 'yauzl'
import * as csvParse from 'csv-parse'

import { InvalidInputFile, UnableToExtract, NoPrimaryColumn } from './errs'
import type * as Gtfs from './gtfsTypes'

import { ipcMain } from 'electron'
import type { BrowserWindow } from 'electron'

// Workaround for https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20497
type PYauzlOpen = (arg1: string, arg2?: yauzl.Options) => Promise<yauzl.ZipFile>

type LoadingStatusStates = 'nofile' | 'error' | 'loading' | 'done'
type LoadingStatusTableState = 'error' | 'loading' | 'done'

export interface LoadingStatus {
  status: LoadingStatusStates
  fileName: string | null
  tables?: Map<string, LoadingStatusTableState>
  error?: Error
}

// Function creating Gtfs.File* from readable streams
async function processToRow (readStream: NodeJS.ReadableStream, tableName: string, primaryKey: string): Promise<Gtfs.FileToRow> {
  const parser = readStream.pipe(csvParse({ bom: true, columns: true }))
  const map: Gtfs.FileToRow = new Map()

  for await (const row of parser) {
    if (row[primaryKey] === undefined) {
      throw new NoPrimaryColumn(`Table ${tableName} is missing its primary column: ${primaryKey}`)
    }

    map.set(row[primaryKey], row)
  }

  return map
}

async function processToList (readStream: NodeJS.ReadableStream, tableName: string, primaryKey: string): Promise<Gtfs.FileToList> {
  const parser = readStream.pipe(csvParse({ bom: true, columns: true }))
  const map: Gtfs.FileToList = new Map()

  for await (const row of parser) {
    if (row[primaryKey] === undefined) {
      throw new NoPrimaryColumn(`Table ${tableName} is missing its primary column: ${primaryKey}`)
    }

    const thisTrip = map.get(row[primaryKey])
    if (thisTrip === undefined) {
      map.set(row[primaryKey], [row])
    } else {
      thisTrip.push(row)
    }
  }

  return map
}

async function processShapes (readStream: NodeJS.ReadableStream, tableName: string, primaryKey: string): Promise<Gtfs.FileShapes> {
  const parser = readStream.pipe(csvParse({ bom: true, columns: true }))
  const rowMap: Gtfs.FileToList = new Map()
  const map: Gtfs.FileShapes = new Map()

  for await (const row of parser) {
    if (row.shape_id === undefined) {
      throw new NoPrimaryColumn(`Table ${tableName} is missing its primary column: ${primaryKey}`)
    }

    const thisTrip = rowMap.get(row.shape_id)
    if (thisTrip === undefined) {
      rowMap.set(row.shape_id, [row])
    } else {
      thisTrip.push(row)
    }
  }

  rowMap.forEach((rows, key) => {
    rowMap.delete(key)

    const points: Array<[number, number]> = rows
      .sort((a, b) => (parseInt(a.shape_pt_sequence) - parseInt(b.shape_pt_sequence)))
      .map(i => [parseFloat(i.shape_pt_lat), parseFloat(i.shape_pt_lon)])

    map.set(key, points)
  })

  return map
}

async function postProcessStops (stopMap: Gtfs.FileToRow | undefined): Promise<Gtfs.StopChildren | undefined> {
  if (stopMap === undefined) { return undefined }
  const childrenMap: Gtfs.StopChildren = new Map()

  for (const [stopId, row] of stopMap) {
    if (row.location_type !== '1' && row.parent_station !== '') {
      const currentChildren = childrenMap.get(row.parent_station)
      if (currentChildren === undefined) {
        childrenMap.set(row.parent_station, [stopId])
      } else {
        currentChildren.push(stopId)
      }
    }
  }

  return childrenMap
}

// Instructions how to handle GTFS file
type Processor<T> = (readStream: NodeJS.ReadableStream, tableName: string, primaryKey: string) => Promise<T>
type FileHandlerDef<T extends keyof Gtfs.Obj> = [T, Processor<Gtfs.Obj[T]>, string]

// file_name → tableName, processingFunction, primaryKey
const fileHandlers: Map<string, FileHandlerDef<keyof Gtfs.Obj>> = new Map([
  ['agency.txt', ['agency', processToRow, 'agency_id']],
  ['stops.txt', ['stops', processToRow, 'stop_id']],
  ['routes.txt', ['routes', processToRow, 'route_id']],
  ['trips.txt', ['trips', processToRow, 'trip_id']],
  ['calendar.txt', ['calendar', processToRow, 'service_id']],
  ['calendar_dates.txt', ['calendarDates', processToList, 'service_id']],
  ['frequencies.txt', ['frequencies', processToList, 'trip_id']],
  ['stop_times.txt', ['stopTimes', processToList, 'trip_id']],
  ['shapes.txt', ['shapes', processShapes, 'shape_id']]
])

// Current state of loading the GTFS
class StatusUpdater {
  window: BrowserWindow
  status: LoadingStatus

  constructor (mainWindow: BrowserWindow) {
    this.window = mainWindow
    this.status = {
      status: 'loading',
      fileName: null
    }
  }

  async send (): Promise<void> {
    this.window.webContents.send('loading-status', this.status)
  }

  setNoFile (): void {
    this.status.status = 'nofile'
  }

  async setError (e: Error): Promise<void> {
    this.status.status = 'error'
    delete this.status.tables
    this.status.error = e

    await this.send()
  }

  async setTableState (table: string, newStatus: LoadingStatusTableState): Promise<void> {
    if (this.status.tables === undefined) { this.status.tables = new Map() }
    this.status.tables.set(table, newStatus)

    await this.send()
  }

  async setDone (): Promise<void> {
    this.status.status = 'done'
    delete this.status.tables
    await this.send()
  }
}

var statusUpdater: StatusUpdater

// Function that process specific input types
async function handleDirectory (directoryPath: string): Promise<Gtfs.Obj> {
  const entryPromises: Array<Promise<void>> = []
  const gtfsObj: Gtfs.Obj = {}

  const handleEntry = async (filename: string): Promise<void> => {
    // Get information what to do with this entry
    const thisFileHandlers = fileHandlers.get(filename)

    if (thisFileHandlers === undefined) { return }
    const [table, processor, primaryKey] = thisFileHandlers

    // Mark this table as loaded
    await statusUpdater.setTableState(table, 'loading')

    const reader = fs.createReadStream(join(directoryPath, filename))

    // @ts-expect-error | no fucking clue
    gtfsObj[table] = await processor(reader, table, primaryKey)
    if (table === 'stops') { gtfsObj._stopChildren = await postProcessStops(gtfsObj.stops) }

    // Mark this table as loaded
    await statusUpdater.setTableState(table, 'done')
  }

  fs.readdirSync(directoryPath).forEach(filename => entryPromises.push(handleEntry(filename)))

  await Promise.all(entryPromises)
  await statusUpdater.setDone()

  return gtfsObj
}

async function handleZip (zipPath: string): Promise<Gtfs.Obj> {
  const entryPromises: Array<Promise<void>> = []
  const gtfsObj: Gtfs.Obj = {}
  let zip: yauzl.ZipFile | undefined

  try {
    zip = await (promisify(yauzl.open) as PYauzlOpen)(zipPath, { autoClose: true })
  } catch {
    // yauzl throws some really nondescriptive errors, so we re-throw our own one
    throw new InvalidInputFile(
            `Provided file: ${zipPath} doesn't point to ` +
            'a directory or a valid zip archive!')
  }

  if (zip === undefined) {
    throw new UnableToExtract(`yauzl was unable to read ZIP file ${zipPath}`)
  }

  const openReadStream = promisify(zip.openReadStream.bind(zip))
  const handleEntry = async (entry: yauzl.Entry): Promise<void> => {
    const thisFileHandler = fileHandlers.get(entry.fileName)

    if (thisFileHandler === undefined) { return }
    const [table, processor, primaryKey] = thisFileHandler
    const reader = await openReadStream(entry)

    await statusUpdater.setTableState(table, 'loading')

    // Check if the input stream was successfully created
    if (reader === undefined) {
      throw new UnableToExtract(
                `yauzl was unable to read file ${entry.fileName} ` +
                'from within provided archive!')
    }

    // @ts-expect-error | no fucking clue
    gtfsObj[table] = await processor(reader, table, primaryKey)
    if (table === 'stops') { gtfsObj._stopChildren = await postProcessStops(gtfsObj.stops) }

    await statusUpdater.setTableState(table, 'done')
  }

  zip.on('entry', entry => {
    entryPromises.push(handleEntry(entry))
  })

  // Wait until all actions have finished
  await pEvent(zip, 'end')
  await Promise.all(entryPromises)
  await statusUpdater.setDone()

  return gtfsObj
}

/**
 * Tries to load whatever the user provided as "GTFS":
 * Either a path to a zip archive, or path to a directory with extracted GTFS files.
 * Then, tries to load GTFS tables into memory.
 */
export async function handleInput (inputPath: string | undefined, mainWin: BrowserWindow): Promise<Gtfs.Obj> {
  statusUpdater = new StatusUpdater(mainWin)

  ipcMain.handle('loading-status-req', async _event => await statusUpdater.send())

  if (inputPath === undefined) {
    statusUpdater.setNoFile()
    throw new InvalidInputFile('No input file provided!')
  } else if (!fs.existsSync(inputPath)) {
    statusUpdater.status.fileName = inputPath
    statusUpdater.setNoFile()
    throw new InvalidInputFile('Input file does not exist!')
  }

  statusUpdater.status.fileName = inputPath
  let gtfsObj: Gtfs.Obj

  try {
    const inputStat = await promisify(fs.stat)(inputPath)

    if (inputStat.isDirectory()) {
      gtfsObj = await handleDirectory(inputPath)
    } else {
      gtfsObj = await handleZip(inputPath)
    }
  } catch (e) {
    await statusUpdater.setError(e)
    throw e
  }

  ipcMain.removeHandler('loading-status-req')
  return gtfsObj
}

File: jvig/gtfs.py
# jvig - GTFS Viewer, created using Flask.
# Copyright © 2022 Mikołaj Kuranowski

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import csv
import logging
import zipfile
from dataclasses import dataclass, field
from datetime import date, timedelta
from io import TextIOWrapper
from math import nan
from pathlib import Path
from typing import IO, Callable, List, Union

from .util import parse_gtfs_date

logger = logging.getLogger("jvig.gtfs")

Row = dict[str, str]
Point = tuple[float, float]

TableToOne = dict[str, Row]
TableToMany = dict[str, list[Row]]
TableToPoints = dict[str, list[Point]]
Table = Union[TableToOne, TableToMany, TableToPoints]


def _get_shape_pt(row: Row) -> Point:
    """Tries to parse a shapes.txt row into a Point tuple.
    If there's anything wrong with the row, returns (nan, nan)."""
    try:
        return float(row.get("shape_pt_lat", nan)), float(row.get("shape_pt_lon", nan))
    except ValueError:
        return nan, nan


_table_keys: dict[str, str] = {
    "agency": "agency_id",
    "stops": "stop_id",
    "routes": "route_id",
    "trips": "trip_id",
    "calendar": "service_id",
    "calendar_dates": "service_id",
    "frequencies": "trip_id",
    "stop_times": "trip_id",
    "shapes": "shape_id",
}


@dataclass
class Gtfs:
    """Gtfs is a class that holds all known GTFS tables.
    It's also responsible for loading the data."""

    agency: TableToOne = field(default_factory=dict)
    stops: TableToOne = field(default_factory=dict)
    stop_children: dict[str, list[str]] = field(default_factory=dict)
    routes: TableToOne = field(default_factory=dict)
    trips: TableToOne = field(default_factory=dict)
    calendar: TableToOne = field(default_factory=dict)
    calendar_dates: TableToMany = field(default_factory=dict)
    frequencies: TableToMany = field(default_factory=dict)
    stop_times: TableToMany = field(default_factory=dict)
    stop_times_by_stops: TableToMany = field(default_factory=dict)
    shapes: TableToPoints = field(default_factory=dict)

    def load_to_row(self, table_name: str, stream: IO[str]) -> None:
        """Loads a table where the key should map into a single row,
        like agency.txt or calendar.txt."""
        primary_key = _table_keys[table_name]
        table: TableToOne = getattr(self, table_name)
        table.clear()

        for row in csv.DictReader(stream):
            # Fix for GTFS feeds without an explicit agency_id
            if table_name in ("agency", "routes") and "agency_id" not in row:
                row["agency_id"] = "(missing)"

            table[row[primary_key]] = row

    def load_to_rows(self, table_name: str, stream: IO[str]) -> None:
        """Loads a table where the key should map into multiple row, like frequencies.txt."""
        primary_key = _table_keys[table_name]
        table: TableToMany = getattr(self, table_name)
        table.clear()

        for row in csv.DictReader(stream):
            table.setdefault(row[primary_key], []).append(row)

    def load_stops(self, table_name: str, stream: IO[str]) -> None:
        """Specialized loader for stops.txt, which loads data into both
        self.stops and self.stop_children."""
        assert table_name == "stops"
        self.stops.clear()
        self.stop_children.clear()

        for row in csv.DictReader(stream):
            self.stops[row["stop_id"]] = row

            # Check if this is a child stop belonging to a larger structure
            parent = row.get("parent_station")
            if parent and row.get("location_type") != "1":
                self.stop_children.setdefault(parent, []).append(row["stop_id"])

    def load_shapes(self, table_name: str, stream: IO[str]) -> None:
        """Specialized loader for shapes.txt to parse the shape."""
        assert table_name == "shapes"
        self.shapes.clear()

        # FIXME: First sort by shape_pt_sequence
        for row in csv.DictReader(stream):
            self.shapes.setdefault(row["shape_id"], []).append(_get_shape_pt(row))

    def load_stop_times(self, table_name: str, stream: IO[str]) -> None:
        """Specialized loader for stop_times.txt, which loads the data
        into self.stop_times and self.stop_times_by_stops."""
        # FIXME: Sort by stop_sequence

        assert table_name == "stop_times"
        self.stop_times.clear()
        self.stop_times_by_stops.clear()

        for row in csv.DictReader(stream):
            self.stop_times.setdefault(row["trip_id"], []).append(row)
            self.stop_times_by_stops.setdefault(row["stop_id"], []).append(row)

    def header_of(self, table_name: str) -> List[str]:
        """Returns the GTFS header of a particlar table."""
        table: Union[TableToOne, TableToMany] = getattr(self, table_name)

        # Get the first entry from the table
        entry = next(iter(table.values()), {})

        # If it's a list of rows, get its first row
        if isinstance(entry, list):
            entry = entry[0]

        # Return the keys of the row
        return list(entry.keys())

    @property
    def _loader_table(self) -> dict[str, Callable[[str, IO[str]], None]]:
        return {
            "agency": self.load_to_row,
            "stops": self.load_stops,
            "routes": self.load_to_row,
            "trips": self.load_to_row,
            "calendar": self.load_to_row,
            "calendar_dates": self.load_to_rows,
            "frequencies": self.load_to_rows,
            "stop_times": self.load_stop_times,
            "shapes": self.load_shapes,
        }

    @classmethod
    def from_directory(cls, where: Path) -> "Gtfs":
        """Loads GTFS data from a directory of .txt files"""
        self = cls()
        loaders = self._loader_table

        for f in where.glob("*.txt"):
            table_name = f.stem
            loader = loaders.get(table_name)

            if loader:
                logger.info(f"Loading table {table_name}")
                with f.open(mode="r", encoding="utf-8-sig", newline="") as stream:
                    loader(table_name, stream)

        return self

    @classmethod
    def from_zip(cls, where: Path) -> "Gtfs":
        """Loads GTFS data from a .zip archive"""
        self = cls()
        loaders = self._loader_table

        with zipfile.ZipFile(where, mode="r") as archive:
            for f in archive.infolist():
                if not f.filename.endswith(".txt"):
                    logger.warn(f"Unrecognized file in zip: {f.filename}")
                    continue

                table_name = f.filename[:-4]
                loader = loaders.get(table_name)

                if loader:
                    logger.info(f"Loading table {table_name}")
                    with archive.open(f, mode="r") as binary_stream:
                        stream = TextIOWrapper(binary_stream, encoding="utf-8-sig", newline="")
                        loader(table_name, stream)

                else:
                    logger.warning(f"Unrecognized file in zip: {f.filename}")

        return self

    @classmethod
    def from_user_input(cls, where: Path) -> "Gtfs":
        """Loads data from a .zip file (if `where` is a file),
        or from a directory with .txt files (if `where` is not a file)"""
        return cls.from_zip(where) if where.is_file() else cls.from_directory(where)

    def all_stops_in_group(self, stop_id: str) -> list[Row]:
        """Returns all stops in the group to which `stop_id` belongs.

        If the stop doesn't exist, returns an empty list.
        If the stop doesn't belong to a group, returns a list with one element.

        In any other cases, the provided stop will always be the
        first element in the returned list.
        """
        if stop_id not in self.stops:
            return []

        stop = self.stops[stop_id]
        stops = [stop]

        if stop.get("location_type") == "1":
            # This stop is the parent station - just append the children
            stops.extend(
                self.stops[child_id]
                for child_id in self.stop_children.get(stop_id, [])
                if child_id in self.stops
            )

        elif stop.get("parent_station") in self.stop_children:
            # This stop is a child in a station - first append the parent,
            # then the other children
            parent_id = stop["parent_station"]

            if parent_id in self.stops:
                stops.append(self.stops[parent_id])

            stops.extend(
                self.stops[child_id]
                for child_id in self.stop_children.get(parent_id, [])
                if child_id in self.stops and child_id != stop_id
            )

        return stops

    def all_dates_of(self, service_id: str) -> set[date]:
        """Returns a set of all date on which a particular calendar is active"""
        dates: set[date] = set()

        # Read the calendar.txt row
        calendar_row = self.calendar.get(service_id)
        if calendar_row:
            active_weekdays = {
                0: calendar_row["monday"] == "1",
                1: calendar_row["tuesday"] == "1",
                2: calendar_row["wednesday"] == "1",
                3: calendar_row["thursday"] == "1",
                4: calendar_row["friday"] == "1",
                5: calendar_row["saturday"] == "1",
                6: calendar_row["sunday"] == "1",
            }
            day = parse_gtfs_date(calendar_row["start_date"])
            end = parse_gtfs_date(calendar_row["end_date"])
            while day <= end:
                if active_weekdays[day.weekday()]:
                    dates.add(day)
                day += timedelta(days=1)

        # Add the calendar_dates.txt rows
        for dates_row in self.calendar_dates.get(service_id, []):
            day = parse_gtfs_date(dates_row["date"])
            if dates_row["exception_type"] == "1":
                dates.add(day)
            elif dates_row["exception_type"] == "2":
                dates.discard(day)

        return dates
